lib/sierra_api_client.rb

```diff
@@ -0,0 +1,146 @@
+require 'net/http'
+require 'net/https'
+require 'uri'
+require 'nypl_log_formatter'
+
+require_relative 'errors'
+require_relative 'sierra_api_response'
+
+class SierraApiClient
+  def initialize(config = {})
+    @config = {
+      base_url: ENV['SIERRA_API_BASE_URL'],
+      client_id: ENV['SIERRA_OAUTH_ID'],
+      client_secret: ENV['SIERRA_OAUTH_SECRET'],
+      oauth_url: ENV['SIERRA_OAUTH_URL'],
+      log_level: 'info'
+    }.merge config
+
+    raise SierraApiClientError.new 'Missing config: neither config.base_url nor ENV.SIERRA_API_BASE_URL are set' unless @config[:base_url]
+    raise SierraApiClientError.new 'Missing config: neither config.client_id nor ENV.SIERRA_OAUTH_ID are set' unless @config[:client_id]
+    raise SierraApiClientError.new 'Missing config: neither config.client_secret nor ENV.SIERRA_OAUTH_SECRET are set ' unless @config[:client_secret]
+    raise SierraApiClientError.new 'Missing config: neither config.oauth_url nor ENV.SIERRA_OAUTH_URL are set ' unless @config[:oauth_url]
+  end
+
+  def get (path, options = {})
+    options = parse_http_options options
+
+    authenticate! if options[:authenticated]
+
+    uri = URI.parse("#{@config[:base_url]}#{path}")
+
+    logger.debug "SierraApiClient: Getting from Sierra api", { uri: uri }
+
+    begin
+      request = Net::HTTP::Get.new(uri)
+
+      # Add bearer token header
+      request["Authorization"] = "Bearer #{@access_token}" if options[:authenticated]
+
+      # Execute request:
+      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme === 'https') do |http|
+        http.request(request)
+      end
+    rescue => e
+      raise SierraApiClientError.new(e), "Failed to GET #{path}: #{e.message}"
+    end
+
+    logger.debug "SierraApiClient: Got Sierra api response", { code: response.code, body: response.body }
+
+    parse_response response
+  end
+
+
+  def post (path, body, options = {})
```

Comment:
Minor inessential comment:

Are the differences in the way the post request and the get request are set up essential? It seems like the only thing that should really be different are 1.  setting up body/content type 2. the actual method.

Reply:
As an end user, I do like having the separate `get` and `post` methods. I guess under the hood these methods could use the same method/share more helper methods? But seems  fine to me. Easy to read.

Reply:
Good notes. I've amended the PR to DRY the GET and POST implementation in a new, generic `do_request` private method, which does most of the work. The prior differences between their implementation were not meaningful.